### PR TITLE
Add omniMAE ViT models

### PR DESCRIPTION
**Pull request by kalyanvasudev**
WIP

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,24 +1,35 @@
...
-[![PWC](https://img.shields.io/endpoint.svg?url=https://paperswithcode.com/badge/omnivore-a-single-model-for-many-visual/semantic-segmentation-on-nyu-depth-v2)](https://paperswithcode.com/sota/semantic-segmentation-on-nyu-depth-v2?p=omnivore-a-single-model-for-many-visual)  
-[![PWC](https://img.shields.io/endpoint.svg?url=https://paperswithcode.com/badge/omnivore-a-single-model-for-many-visual/scene-recognition-on-sun-rgbd)](https://paperswithcode.com/sota/scene-recognition-on-sun-rgbd?p=omnivore-a-single-model-for-many-visual)  
-[![PWC](https://img.shields.io/endpoint.svg?url=https://paperswithcode.com/badge/omnivore-a-single-model-for-many-visual/image-classification-on-inaturalist-2018)](https://paperswithcode.com/sota/image-classification-on-inaturalist-2018?p=omnivore-a-single-model-for-many-visual)  
-[![PWC](https://img.shields.io/endpoint.svg?url=https://paperswithcode.com/badge/omnivore-a-single-model-for-many-visual/action-recognition-in-videos-on-something)](https://paperswithcode.com/sota/action-recognition-in-videos-on-something?p=omnivore-a-single-model-for-many-visual)
+This repository contains PyTorch evaluation code, pretrained models for the following papers:
```
Comment (mannatsingh) [inline README.md:3]: nit: the model zoo, PyTorch evaluation code, and tutorials

Are you planning on adding eval code?

File: omnimae_README.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
+# Omnivore: A Single Model for Many Visual Modalities
```
Comment (mannatsingh) [inline omnimae_README.md:1]: nit for file names - let's use README as the prefix? so README_omnimae.md

File: omnimae_README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  booktitle={TODO},
+  year={2022}
+}
+```
+
+## Contributing
+We welcome your pull requests! Please see [CONTRIBUTING](CONTRIBUTING.md) and [CODE_OF_CONDUCT](CODE_OF_CONDUCT.md) for more information.
+
+## License
+Omnivore is released under the CC-BY-NC 4.0 license. See [LICENSE](LICENSE) for additional details. However the Swin Transformer implementation is additionally licensed under the Apache 2.0 license (see [NOTICE](NOTICE) for additional details).
```
Comment (mannatsingh) [inline omnimae_README.md:35]: Change the text here

File: omnimae/vision_transformer.py [hunk 4]
```diff
@@ -0,0 +1,1121 @@
...
+
+# Code modified from 
+# https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py ;
+# https://github.com/facebookresearch/deit/blob/main/models.py 
+# and https://github.com/facebookresearch/vissl/blob/main/vissl/models/trunks/vision_transformer.py
+# and is licensed under the license found in the
+# NOTICE file in the root directory of this source tree.
+
+
+# FIXME: DO NOT OPEN SOURCE THIS - we're missing proper attributions!
```
Comment (mannatsingh) [inline omnimae/vision_transformer.py:16]: need to remove this line

also, we need to copy any copyright headers from the sources we modified from

File: omnimae/vision_transformer.py [hunk 5]
```diff
@@ -0,0 +1,1121 @@
...
+        )
+        self.layer_scale_type = layer_scale_type
+
+        # LVViT non-skip. Can be thought of as a special case of layer_scale
+        if non_skip_wt_learnable is False:
+            self.non_skip_wt = non_skip_wt
+        else:
+            self.non_skip_wt = nn.Parameter((torch.ones(1) * non_skip_wt).squeeze())
+
+        # Layerscale
```
Comment (mannatsingh) [inline omnimae/vision_transformer.py:224]: I think we can remove some of these unused bits - but can also remove them later, no big deal

Comment (mannatsingh) [inline omnimae/vision_transformer.py:224]: I think we can remove some of these unused bits - but can also remove them later, no big deal

File: README.md [hunk 6]
```diff
@@ -1,24 +1,35 @@
...
+  author={Girdhar, Rohit and El-Nouby Alaa and Singh, Mannat and Alwala, Kalyan Vasudev and Joulin, Armand and Misra, Ishan},
+  booktitle={TODO},
+  year={2022}
+}
+```
+</details>
 
-## Usage
 
 ### Setup and Installation
```
Comment (mannatsingh) [inline README.md:34]: I think we can remove the setup and installation intructions here?

File: omnimae_README.md (hunk 2 above)
Comment (mannatsingh) [reply omnimae_README.md:1]: Also need to fix the title

File: omnimae/vision_transformer.py [hunk 7]
```diff
@@ -0,0 +1,1121 @@
...
+        super().__init__()
+        self.time_dim = time_dim
+        assert ntimes > 0
+        assert pad_type in ["zero", "repeat"]
+        self.ntimes = ntimes
+        self.pad_type = pad_type
+
+    def forward(self, x):
+        if x.ndim == 4:
+            # B, C, H, W -> B, C, T, H, W
```
Comment (sayakpaul) [inline omnimae/vision_transformer.py:65]: Shouldn't the time axis precede the channel axis? `B, T, C, H, W` instead of `B, C, T, H, W`?

File: README.md [hunk 8]
```diff
@@ -1,120 +1,34 @@
-# Omnivore: A Single Model for Many Visual Modalities
+# Omnivorous model architectures for Image and Video classfication and SSL
```
Comment (mannatsingh) [inline README.md:1]: Omnivorous modeling for visual modalities

**Review by mannatsingh — APPROVED** [review]
Will torchhub come as a follow up PR? If so, this looks good to merge once you fix the minor nits

File: omnimae/vision_transformer.py (hunk 7 above)
Comment (mannatsingh) [reply omnimae/vision_transformer.py:65]: Sorry I didn't get your question. The code just adds a new time dimension, there shouldn't be any order swapping.

Comment (sayakpaul) [reply omnimae/vision_transformer.py:65]: I meant why does the time axis precede the channel axis? At least that's what I understood from this comment: `# B, C, H, W -> B, C, T, H, W`.

Comment (mannatsingh) [reply omnimae/vision_transformer.py:65]: The time axis doesn't precede the channel axis - this function takes an image which has no time axis and adds a time axis (converts an image to a video).

That is the same setup as our omnivore paper which you can play around with in https://github.com/facebookresearch/omnivore/blob/main/inference_tutorial.ipynb.

Comment (sayakpaul) [reply omnimae/vision_transformer.py:65]: I understand that. I am simply asking why it's added after the channel axis. Why couldn't it be `B, T, C, H, W`?

Comment (mannatsingh) [reply omnimae/vision_transformer.py:65]: That's what I had mentioned earlier :) 

> Actually, we keep the time and spatial axes together so that they can be patchified together

To create a spatiotemporal patch you need the time and spatial axes to be next to each other.

Or if I'm not able to explain it I'll leave it to @kalyanvasudev to answer your question!

Comment (sayakpaul) [reply omnimae/vision_transformer.py:65]: Ah I now get it. Thank you for being patient. I apologize for not being attentive enough.

Comment (mannatsingh) [reply omnimae/vision_transformer.py:65]: No worries, glad I could help clarify things!

File: NOTICE [hunk 9]
```diff
@@ -204,3 +204,211 @@ Copyright 2018-2019 Open-MMLab. All rights reserved.
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
+   
+=======================================================================
+Vision Transformer implementation's Apache 2.0 license
```
[the remark this answers is not in the document]
Comment (kalyanvasudev) [reply NOTICE:209]: Yes, its from timm. Let me add that.

**Comment by kalyanvasudev** [comment]
@mannatsingh, seems like I dont have the write access. Feel free to merge this PR if everything looks good. Thanks!
